
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>CanSig: discovering cancer signatures &#8212; CanSig  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="_static/styles/bootstrap.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=796348d33e8b1d947c94" rel="stylesheet">
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94">
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'index';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Preprocessing" href="preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fa-solid fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="#">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">CanSig  documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="preprocessing.html">
                        Preprocessing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="pipeline-advanced.html">
                        Advanced pipeline usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="formatting-data.html">
                        Input data format
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="cansig-for-python-coders.html">
                        CanSig as a package
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="troubleshooting.html">
                        Troubleshooting
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="api.html">
                        API
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      
        <div class="navbar-end-item navbar-persistent--container">
          
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
        </div>
      
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://github.com/BoevaLab/CanSig" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fab fa-github-square"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
      </ul>
      </div>
      
    </div>
  </div>


  
  
    <div class="navbar-persistent--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
    </div>
  

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="preprocessing.html">
                        Preprocessing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="pipeline-advanced.html">
                        Advanced pipeline usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="formatting-data.html">
                        Input data format
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="cansig-for-python-coders.html">
                        CanSig as a package
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="troubleshooting.html">
                        Troubleshooting
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="contributing.html">
                        Contributing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="api.html">
                        API
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://github.com/BoevaLab/CanSig" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fab fa-github-square"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
      </ul>
      </div>
      
    </div>
    
  </div>

  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

  
  <div id="rtd-footer-container"></div>

      </div>
      <main id="main-content" class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <section id="cansig-discovering-cancer-signatures">
<h1>CanSig: discovering cancer signatures<a class="headerlink" href="#cansig-discovering-cancer-signatures" title="Permalink to this heading">#</a></h1>
<p>Human tumors are highly heterogeneous in their cell composition; specifically, they exhibit heterogeneity in transcriptional states of malignant cells, as has been recently discovered through single-cell RNA sequencing (scRNA-seq). Distinct states of malignant cells have been linked to variability in tumorigenic properties and resistance to anti-cancer treatment. Despite the fact that scRNA-seq data contain all necessary information to uncover shared transcriptional states of malignant cells in tumors, jointly analyzing cells from multiple cancer patients comes with its set of challenges including batch correction and accounting for patient-specific genetic background driving differences between gene expression vectors. We propose CanSig, an easy-to-use approach designed to discover known and de novo shared signatures in cancer single cells. CanSig preprocesses, integrates and analyzes scRNA-seq data to provide new signatures of shared transcriptional states and links these states to known pathways.</p>
<a class="reference internal image-reference" href="_images/pipeline.png"><img alt="The diagram representing CanSig pipeline." class="align-center" src="_images/pipeline.png" style="width: 600px;" /></a>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A preprint describing the pipeline and case studies is <a class="reference external" href="https://doi.org/10.1101/2022.04.14.488324">now available</a>.</p>
</div>
<section id="navigation">
<h2>Navigation<a class="headerlink" href="#navigation" title="Permalink to this heading">#</a></h2>
<div class="toctree-wrapper compound">
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="preprocessing.html">Preprocessing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#loading-the-data">Loading the data</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#id4">Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#outputs">Outputs</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pipeline-advanced.html">Advanced pipeline usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#using-the-plotting-utilities">Using the plotting utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#saving-signatures-and-scoring-cells">Saving signatures and scoring cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#running-differential-cnv-analysis">Running differential CNV analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#using-custom-models">Using custom models</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#understanding-all-the-pipeline-command-line-options">Understanding all the pipeline command line options</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="formatting-data.html">Input data format</a></li>
<li class="toctree-l1"><a class="reference internal" href="cansig-for-python-coders.html">CanSig as a package</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting</a><ul>
<li class="toctree-l2"><a class="reference internal" href="troubleshooting.html#gsea-does-not-run">GSEA does not run</a></li>
<li class="toctree-l2"><a class="reference internal" href="troubleshooting.html#differential-cnv-analysis-does-not-work">Differential CNV analysis does not work</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#submitting-an-issue">Submitting an issue</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#submitting-a-pull-request">Submitting a Pull Request</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#for-new-developers">For new developers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>
</div>
<ul class="simple">
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>
<section id="getting-started">
<h2>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this heading">#</a></h2>
<section id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Permalink to this heading">#</a></h3>
<p>This tutorial assumes you have access to a Unix terminal (e.g., Linux, BSD, MacOS). If you use Windows, install <a class="reference external" href="https://docs.microsoft.com/en-us/windows/wsl/install">Windows Subsystem for Linux</a> first.</p>
<p>When the Unix terminal is ready, you will need to <a class="reference external" href="https://wiki.python.org/moin/BeginnersGuide">install Python</a>. One popular solution is <a class="reference external" href="https://www.anaconda.com/">Anaconda</a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>CanSig requires Python 3.8 (or a most recent version).</p>
</div>
<p>Install the package:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>cansig
</pre></div>
</div>
<p>Alternatively, in the <a class="reference external" href="https://github.com/boevalab/cansig">repository</a> we provide an <code class="docutils literal notranslate"><span class="pre">environment.yml</span></code> file <a class="reference external" href="https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file">which can be used to create a new conda environment</a>.</p>
</section>
<section id="example-dataset">
<h3>Example dataset<a class="headerlink" href="#example-dataset" title="Permalink to this heading">#</a></h3>
<p>We will download the example data set using an auxiliary script:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>python<span class="w"> </span>-m<span class="w"> </span>cansig.run.download
</pre></div>
</div>
<p>You should see a new directory <code class="docutils literal notranslate"><span class="pre">data</span></code>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>ls<span class="w"> </span>data/tutorial/simulated
malignant.h5ad<span class="w">  </span>msigdb74.gmt<span class="w">  </span>README.md
</pre></div>
</div>
<p>This directory contains a H5AD file with <em>simulated</em> malignant cells and a GMT file with gene signatures. More information about these are in the README file.</p>
<p>For convenience we will <a class="reference external" href="https://linuxhint.com/variables_bash/">define variables</a> storing the locations:</p>
<p>Let’s take a look at this data set. Note that T. Chari and L. Pachter in <a class="reference external" href="https://doi.org/10.1101/2021.08.25.457696">The Specious Art of Single-Cell Genomics</a> demonstrate that single-cell data is hard to visualise and popular dimension reduction strategies may lead to spurious conclusions.</p>
<p>We will use UMAP for demonstratory purposes to demonstrate the signals in the data we simulated. We suggest to install and run <a class="reference external" href="https://jupyter.org/">Jupyter Notebook</a> for this.</p>
<p>Once the notebook is running, we can load the data:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;data/tutorial/simulated/malignant.h5ad&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span> <span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="s2">&quot;subclonal&quot;</span><span class="p">],</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/tutorial-simulated-umap.png"><img alt="UMAP visualisation of simulated cells." class="align-center" src="_images/tutorial-simulated-umap.png" style="width: 400px;" /></a>
<p>We simulated samples coming from different patients, with different subclonal composition and different malignant cell types.
Different patients may be missing some of the cell types. Let’s visualise how many cells with particular cell type are within each sample:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">plot_joint</span><span class="p">(</span><span class="n">col1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">col2</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
   <span class="k">return</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">data</span>
                      <span class="o">.</span><span class="n">obs</span>
                      <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">col1</span><span class="p">,</span> <span class="n">col2</span><span class="p">])</span>
                      <span class="o">.</span><span class="n">size</span><span class="p">()</span>
                      <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                      <span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">col1</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">col2</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
                      <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                      <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;g&quot;</span><span class="p">,</span>
                     <span class="p">)</span>

<span class="n">plot_joint</span><span class="p">(</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span> <span class="s2">&quot;celltype&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/tutorial-simulated-joint.png"><img alt="Joint (empirical) distribution of cell types and patients." class="align-center" src="_images/tutorial-simulated-joint.png" style="width: 400px;" /></a>
</section>
<section id="running-the-analysis">
<h3>Running the analysis<a class="headerlink" href="#running-the-analysis" title="Permalink to this heading">#</a></h3>
<p>As we know what we can expect in this data set, we will run the analysis:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>python<span class="w"> </span>-m<span class="w"> </span>cansig.run.pipeline<span class="w"> </span>data/tutorial/simulated/malignant.h5ad
<span class="w">                                </span>--gene-sets<span class="w"> </span>data/tutorial/simulated/msigdb74.gmt
<span class="w">                                </span>--output<span class="w"> </span>tutorial-output
<span class="w">                                </span>--batch<span class="w"> </span>sample_id
<span class="w">                                </span>--n-top-genes<span class="w"> </span><span class="m">2000</span>
<span class="w">                                </span>--dimensions<span class="w"> </span><span class="m">4</span><span class="w"> </span><span class="m">6</span><span class="w"> </span><span class="m">8</span>
<span class="w">                                </span>--model-runs<span class="w"> </span><span class="m">1</span>
<span class="w">                                </span>--clusters<span class="w"> </span><span class="m">4</span><span class="w"> </span><span class="m">6</span><span class="w"> </span><span class="m">8</span><span class="w"> </span><span class="m">10</span>
<span class="w">                                </span>--cluster-runs<span class="w"> </span><span class="m">2</span>
<span class="w">                                </span>--max-epochs<span class="w"> </span><span class="m">200</span>
</pre></div>
</div>
<p>Apart from specifying the data and gene sets location we specified the output path <code class="docutils literal notranslate"><span class="pre">tutorial-output</span></code> and the column with experimental batch variable <code class="docutils literal notranslate"><span class="pre">sample_id</span></code>, as CanSig will try to remove batch effects using <a class="reference external" href="https://scvi-tools.org/">scVI</a>. We will do the data integration step for three latent space dimensionalities (<code class="docutils literal notranslate"><span class="pre">--dimensions</span> <span class="pre">4</span> <span class="pre">6</span> <span class="pre">8</span></code>). We will train each of these models once (<code class="docutils literal notranslate"><span class="pre">--model-runs</span> <span class="pre">1</span></code>), although one could also ablate the influence to random seed choice by increasing this value. Each model will be trained for maximally 200 epochs.</p>
<p>We will also do the Leiden clustering into 4, 6, 8, and 10 clusters. For each number of clusters we will run the Leiden clustering twice (<code class="docutils literal notranslate"><span class="pre">--cluster-runs</span> <span class="pre">2</span></code>) as this is a randomised algorithm.</p>
<p>This step should take up to thirty minutes on a computer with modern GPU. For computers with access only to CPU it is possible to run the pipeline for a smaller number of models or training them for shorter time, although this may affect the results.</p>
<p>Additional options are visible by using the <code class="docutils literal notranslate"><span class="pre">--help</span></code> flag:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case of additional continuous or discrete variables, which may confound the expression (e.g., the sequencing technology used), one could specify it via the <code class="docutils literal notranslate"><span class="pre">--continuous-covariates</span></code> and <code class="docutils literal notranslate"><span class="pre">--discrete-covariates</span></code> flags.</p>
<p>For more information, run</p>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">-m</span> <span class="pre">cansig.run.pipeline</span> <span class="pre">--help</span></code></p>
</div>
<p>After the command has finished running, the results will be available in the generated directory <code class="docutils literal notranslate"><span class="pre">tutorial-output/</span></code>.</p>
<p>Let’s analyse its structure.</p>
<ul class="simple">
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">integration/</span></code>: each integration model has a separate directory representing the latent representations inferred by the batch correction/dimension reduction method.</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/integration-settings.json</span></code>: summary of the parameters used for the integration method.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/latent-representations.csv</span></code>: coordinates in the latent space inferred by the integration method.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">postprocessing/</span></code>: each clustering configuration on each model has a separate directory with the postprocessing information saved.</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/cluster-labels.csv</span></code>: cluster assignments for this run .</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/cluster-settings.json</span></code>: summary of the parameters used for the clustering step.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/gsea-dataframe.csv</span></code>: results of GSEA run in this setting (full table, including non-significant pathways and negative enrichment score pathways).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/gsea-settings.json</span></code>: summary of the parameters used for the GSEA run.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/integration-settings.json</span></code>: summary of the parameters used for the integration method.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/latent-space-dimred.png</span></code>: scatter plot of the PCA on the latent space colored according to batch and clustering.</p></li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">{rundir}/signatures/</span></code>: directory containing the de novo signatures as the results of the differential gene expression analysis.</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">signature_cl{CLUSTER}.csv</span></code>: results of the differential gene expression analysis for cluster CLUSTER, ranked according to z-score.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">metasignatures/</span></code>: summary of the runs obtained by clustering the signatures from different postprocessing directories.</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">signatures/</span></code>: contains the files with the clustered signatures (“meta-signatures”).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cell-metamembership.csv</span></code>: cells assigned to their meta-signatures.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">similarity-matrix.csv</span></code>: similarity between different postprocessing runs, used to create meta-signatures.</p></li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">figures/</span></code>: generated figures</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">heatmap-metasignatures.png</span></code>: visualisation of the similarity between different postprocessing runs, used to create meta-signatures.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">clustermap-metasignatures-correlation.png</span></code>: correlation between cell scores obtained via scoring with meta-signatures.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
<p>For example, the similarity between used to create metasignatures in our case looks in the following manner:</p>
<a class="reference internal image-reference" href="_images/tutorial-heatmap-metasignatures.png"><img alt="Similarity heatmap." class="align-center" src="_images/tutorial-heatmap-metasignatures.png" style="width: 400px;" /></a>
<p>so one can hope that there are three different meta-signatures. When the cells are scored according to them the following correlation matrix is obtained:</p>
<a class="reference internal image-reference" href="_images/tutorial-clustermap-metasignatures-correlation.png"><img alt="Meta-signatures scores correlation matrix." class="align-center" src="_images/tutorial-clustermap-metasignatures-correlation.png" style="width: 400px;" /></a>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Each directory in <code class="docutils literal notranslate"><span class="pre">postprocessing/</span></code> is a result of running integration method, clustering, and gene set enrichment analysis. We ablate the hyperparameters and  in <code class="docutils literal notranslate"><span class="pre">metasignatures/</span></code> we assemble this information to hope that the aggregated results will be more robust to changes in hyperparameters. Note that there are many possible ways of doing it and perfectly one would apply resampling techniques, such as bootstrap, to understand whether the final output is robust enough.
We advise to use CanSig with this limitation in mind and we recommend <a class="reference external" href="https://www.fharrell.com/post/badb/index.html">F. Harrell’s blogpost</a> for an excellent overview of robust biomarker research.</p>
</div>
</section>
<section id="comparing-with-the-ground-truth">
<h3>Comparing with the ground truth<a class="headerlink" href="#comparing-with-the-ground-truth" title="Permalink to this heading">#</a></h3>
<p>As we mentioned above, unsupervised learning and scientific discovery are hard problems and although CanSig can be used for exploratory data analysis, it is <em>not</em> a replacement for additional validation (by means of putting the findings in the wider biological context, using appropriate bootstrapping technique, and designing wet-lab interventional studies).
We can however see how well the found metasignatures correspond to the cell types in the simulated data.</p>
<p>Let’s open Jupyter and run the following code:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>


<span class="k">def</span> <span class="nf">plot_joint</span><span class="p">(</span><span class="n">col1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">col2</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">data</span>
                       <span class="o">.</span><span class="n">obs</span>
                       <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">col1</span><span class="p">,</span> <span class="n">col2</span><span class="p">])</span>
                       <span class="o">.</span><span class="n">size</span><span class="p">()</span>
                       <span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
                       <span class="o">.</span><span class="n">pivot</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">col1</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">col2</span><span class="p">,</span> <span class="n">values</span><span class="o">=</span><span class="mi">0</span><span class="p">),</span>
                       <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                       <span class="n">fmt</span><span class="o">=</span><span class="s2">&quot;g&quot;</span><span class="p">,</span>
                      <span class="p">)</span>

<span class="c1"># Load the simulated cells</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;data/tutorial/simulated/malignant.h5ad&quot;</span><span class="p">)</span>

<span class="c1"># Load the meta-signatures</span>
<span class="n">metamembership</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;tutorial-output/metasignatures/cell-metamembership.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Add the meta-signatures membership to the data</span>
<span class="n">data</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;metamembership&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">metamembership</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">data</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>


<span class="n">plot_joint</span><span class="p">(</span><span class="s2">&quot;sample_id&quot;</span><span class="p">,</span> <span class="s2">&quot;metamembership&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/tutorial-joint-metasignature-patient.png"><img alt="Joint (empirical) distribution of meta-signature assignment and patient." class="align-center" src="_images/tutorial-joint-metasignature-patient.png" style="width: 400px;" /></a>
<p>As we can see, the cells have been decomposed into three meta-signatures as well as undetermined cells.</p>
<p>We can also visualise the (empirical) joint distribution of ground truth cell types and assigned memberships to meta-signatures:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">plot_joint</span><span class="p">(</span><span class="s2">&quot;celltype&quot;</span><span class="p">,</span> <span class="s2">&quot;metamembership&quot;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/tutorial-joint-metasignature-celltype.png"><img alt="Joint (empirical) distribution of meta-signature assignment and ground truth." class="align-center" src="_images/tutorial-joint-metasignature-celltype.png" style="width: 400px;" /></a>
<p>It seems that <code class="docutils literal notranslate"><span class="pre">metasig1</span></code> is considerably smaller than the other meta-signatures as well as the set of undetermined cells. We see that for both types <code class="docutils literal notranslate"><span class="pre">metasig2</span></code> or <code class="docutils literal notranslate"><span class="pre">metasig3</span></code> seem to be mostly capturing <code class="docutils literal notranslate"><span class="pre">program2</span></code> and <code class="docutils literal notranslate"><span class="pre">program3</span></code>.</p>
</section>
</section>
<section id="tutorials">
<h2>Tutorials<a class="headerlink" href="#tutorials" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>To learn more about the command line pipeline, see <a class="reference internal" href="pipeline-advanced.html#pipeline-advanced"><span class="std std-ref">Advanced pipeline usage</span></a>.</p></li>
<li><p>To learn how to run CanSig elements in a Python script or a Jupyter notebook, see <a class="reference internal" href="cansig-for-python-coders.html#coders"><span class="std std-ref">CanSig as a package</span></a>.</p></li>
<li><p>To learn about the preprocessing module (used to prepare the raw data into the .h5ad format), see the <a class="reference internal" href="preprocessing.html#preprocessing"><span class="std std-ref">Preprocessing</span></a>.</p></li>
<li><p>To learn how to format already preprocessed .csv files into .h5ad files, see <a class="reference internal" href="formatting-data.html#formatting"><span class="std std-ref">Input data format</span></a>.</p></li>
<li><p>To ensure smooth running of CanSig, check out <a class="reference internal" href="troubleshooting.html#troubleshooting"><span class="std std-ref">the troubleshooting page</span></a>.</p></li>
</ul>
</section>
<section id="contributing">
<h2>Contributing<a class="headerlink" href="#contributing" title="Permalink to this heading">#</a></h2>
<p>For the contribution guide and instructions for new developers, see <a class="reference internal" href="contributing.html#contribution-guide"><span class="std std-ref">Contributing</span></a>.</p>
</section>
</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='right-next' id="next-link" href="preprocessing.html" title="next page">
  <div class="prev-next-info">
      <p class="prev-next-subtitle">next</p>
      <p class="prev-next-title">Preprocessing</p>
  </div>
  <i class="fa-solid fa-angle-right"></i>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
<div class="tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
</div>
<nav id="bd-toc-nav" class="page-toc">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#navigation">
   Navigation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#getting-started">
   Getting started
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#installation">
     Installation
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example-dataset">
     Example dataset
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#running-the-analysis">
     Running the analysis
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#comparing-with-the-ground-truth">
     Comparing with the ground truth
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tutorials">
   Tutorials
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributing">
   Contributing
  </a>
 </li>
</ul>

</nav>
</div>

<div class="toc-item">
  
<div id="searchbox"></div>
</div>

<div class="toc-item">
  
</div>

<div class="toc-item">
  
<div class="tocsection sourcelink">
    <a href="_sources/index.rst.txt">
        <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
</div>

</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2022 CanSig team.<br>

</p>

  </div>
  
  <div class="footer-item">
    <p class="theme-version">
    Built with the
    <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">
        PyData Sphinx Theme
    </a>
    0.12.0.
</p>
  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.3.0.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>