
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>CanSig: discovering cancer signatures &#8212; CanSig  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="_static/styles/bootstrap.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">

  
  <link href="_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=796348d33e8b1d947c94" rel="stylesheet">
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94">
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'index';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Preprocessing" href="preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fa-solid fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="#">

  
  
  
  
  
  
  

  
  
    <p class="title logo__title">CanSig  documentation</p>
  
</a>
    
  </div>

  
  <div class="col-lg-9 navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="preprocessing.html">
                        Preprocessing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="pipeline-advanced.html">
                        Advanced pipeline usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="interpretation.html">
                        Data interpretation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="formatting-data.html">
                        Input data format
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="cansig-for-python-coders.html">
                        CanSig for Pythonists
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="troubleshooting.html">
                        Troubleshooting / Checklist
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="contributing.html">
                        Contributing
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      
        <div class="navbar-end-item navbar-persistent--container">
          
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
        </div>
      
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://github.com/BoevaLab/CanSig" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fab fa-github-square"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
      </ul>
      </div>
      
    </div>
  </div>


  
  
    <div class="navbar-persistent--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
    </div>
  

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar hide-on-wide">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="preprocessing.html">
                        Preprocessing
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="pipeline-advanced.html">
                        Advanced pipeline usage
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="interpretation.html">
                        Data interpretation
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="formatting-data.html">
                        Input data format
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="cansig-for-python-coders.html">
                        CanSig for Pythonists
                      </a>
                    </li>
                
            <div class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    More
                </button>
                <div class="dropdown-menu">
                    
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="troubleshooting.html">
                        Troubleshooting / Checklist
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="contributing.html">
                        Contributing
                      </a>
                    </li>
                
                </div>
            </div>
            
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://github.com/BoevaLab/CanSig" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fab fa-github-square"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
      </ul>
      </div>
      
    </div>
    
  </div>

  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

  
  <div id="rtd-footer-container"></div>

      </div>
      <main id="main-content" class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  <section id="cansig-discovering-cancer-signatures">
<h1>CanSig: discovering cancer signatures<a class="headerlink" href="#cansig-discovering-cancer-signatures" title="Permalink to this heading">#</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This package is currently in experimental stage. Changes in the API may appear.</p>
</div>
<p>Human tumors are highly heterogeneous in their cell composition; specifically, they exhibit heterogeneity in transcriptional states of malignant cells, as has been recently discovered through single-cell RNA sequencing (scRNA-seq). Distinct states of malignant cells have been linked to variability in tumorigenic properties and resistance to anti-cancer treatment. Despite the fact that scRNA-seq data contain all necessary information to uncover shared transcriptional states of malignant cells in tumors, jointly analyzing cells from multiple cancer patients comes with its set of challenges including batch correction and accounting for patient-specific genetic background driving differences between gene expression vectors. We propose CanSig, an easy-to-use approach designed to discover known and de novo shared signatures in cancer single cells. CanSig preprocesses, integrates and analyzes scRNA-seq data to provide new signatures of shared transcriptional states and links these states to known pathways.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Add a picture of the pipeline here with the figure caption as in the preprint.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A preprint describing the pipeline and case studies is <a class="reference external" href="https://doi.org/10.1101/2022.04.14.488324">now available</a>.</p>
</div>
<section id="navigation">
<h2>Navigation<a class="headerlink" href="#navigation" title="Permalink to this heading">#</a></h2>
<div class="toctree-wrapper compound">
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="preprocessing.html">Preprocessing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#loading-the-data">Loading the data</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#id4">Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#outputs">Outputs</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pipeline-advanced.html">Advanced pipeline usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#using-the-plotting-utilities">Using the plotting utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#saving-signatures-and-scoring-cells">Saving signatures and scoring cells</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#running-differential-cnv-analysis">Running differential CNV analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#using-custom-models">Using custom models</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#understanding-all-the-pipeline-command-line-options">Understanding all the pipeline command line options</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="interpretation.html">Data interpretation</a></li>
<li class="toctree-l1"><a class="reference internal" href="formatting-data.html">Input data format</a></li>
<li class="toctree-l1"><a class="reference internal" href="cansig-for-python-coders.html">CanSig for Pythonists</a></li>
<li class="toctree-l1"><a class="reference internal" href="troubleshooting.html">Troubleshooting / Checklist</a><ul>
<li class="toctree-l2"><a class="reference internal" href="troubleshooting.html#checklist">Checklist</a></li>
<li class="toctree-l2"><a class="reference internal" href="troubleshooting.html#id4">Troubleshooting</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#submitting-an-issue">Submitting an issue</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#submitting-a-pull-request">Submitting a Pull Request</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#for-new-developers">For new developers</a></li>
</ul>
</li>
</ul>
</div>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>
<section id="getting-started">
<h2>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this heading">#</a></h2>
<p>This tutorial assumes you have access to a Unix terminal (e.g., Linux, BSD, MacOS). If you use Windows, install <a class="reference external" href="https://docs.microsoft.com/en-us/windows/wsl/install">Windows Subsystem for Linux</a> first.</p>
<p>When the Unix terminal is ready, you will need to <a class="reference external" href="https://wiki.python.org/moin/BeginnersGuide">install Python</a>. One popular solution is <a class="reference external" href="https://www.anaconda.com/">Anaconda</a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>CanSig requires Python 3.8 (or a most recent version).</p>
</div>
<section id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Permalink to this heading">#</a></h3>
<p>Install the package:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ pip install cansig
</pre></div>
</div>
</section>
<section id="example-dataset">
<h3>Example dataset<a class="headerlink" href="#example-dataset" title="Permalink to this heading">#</a></h3>
<p>To demonstrate the pipeline usage, we will analyse a small esophageal dataset, which has been already preprocessed (for more information about the dataset and the data curation, see <a class="reference internal" href="preprocessing.html#preprocessing"><span class="std std-ref">Preprocessing</span></a>).</p>
<p>We will download it using an auxiliary script:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m cansig.run.download PIPELINE
</pre></div>
</div>
<p>You should see a new directory <code class="docutils literal notranslate"><span class="pre">data</span></code> with a file <code class="docutils literal notranslate"><span class="pre">pipeline-tutorial.h5ad</span></code>.</p>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>Add the description of this dataset and UMAP visualisations.</p>
</div>
</section>
<section id="running-the-analysis">
<h3>Running the analysis<a class="headerlink" href="#running-the-analysis" title="Permalink to this heading">#</a></h3>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you are familiar with Python, you may prefer running the analysis in a custom manner. See the <a class="reference internal" href="pipeline-advanced.html#pipeline-advanced"><span class="std std-ref">Advanced pipeline usage</span></a> and <a class="reference internal" href="cansig-for-python-coders.html#coders"><span class="std std-ref">Cansig for python coders</span></a>.</p>
</div>
<p>We will run the analysis pipeline with a single command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m cansig.run.pipeline data/pipeline-tutorial/data.hdf5
                                --batch batch
                                --gene-sets MSigDB_Hallmark_2020 <span class="se">\</span>
                                --dimensions <span class="m">4</span> <span class="m">6</span> --model-runs <span class="m">1</span> <span class="se">\</span>
                                --clusters <span class="m">2</span> <span class="m">3</span> <span class="m">5</span>  --cluster-runs <span class="m">1</span> <span class="se">\</span>
                                --output tutorial-output
</pre></div>
</div>
<p>This command launched the training of several models, used for data integration (batch correction and dimension reduction).
We marked that the batch column in the dataset is named <code class="docutils literal notranslate"><span class="pre">batch</span></code> and we will reduce the dimension either to 4 or 6 (<code class="docutils literal notranslate"><span class="pre">--dimensions</span> <span class="pre">4</span> <span class="pre">6</span></code>).
For each of these dimensionalities, we do a single run (<code class="docutils literal notranslate"><span class="pre">--model-runs</span> <span class="pre">1</span></code>). Increasing this number will train more models, differing in random seeds.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case of additional continuous or discrete variables, which may confound the expression (e.g., the sequencing technology used), one could specify it via the <code class="docutils literal notranslate"><span class="pre">--continuous-covariates</span></code> and <code class="docutils literal notranslate"><span class="pre">--discrete-covariates</span></code> flags.</p>
<p>For more information, run</p>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">-m</span> <span class="pre">cansig.run.pipeline</span> <span class="pre">--help</span></code></p>
</div>
<p>Then, for each of the four models, we apply clustering into 2, 3, or 5 communities (<code class="docutils literal notranslate"><span class="pre">--clusters</span> <span class="pre">2</span> <span class="pre">3</span> <span class="pre">5</span></code>).
Again, one can repeat this steps, using different random seeds, by increasing <code class="docutils literal notranslate"><span class="pre">--cluster-runs</span></code>.</p>
<p>We specified the database to be used (<code class="docutils literal notranslate"><span class="pre">--gene-sets</span> <span class="pre">MSigDB_Hallmark_2020</span></code>).
This is one of many databases available online. Alternatively, one can provide a path to the GMT file.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<blockquote>
<div><p>Running the pipeline in this way requires internet connectivity.
If not connected to the internet, to run the same analysis, you need to download the file <code class="docutils literal notranslate"><span class="pre">h.all.v{current_version}.symbols.gmt</span></code>, go to the directory where this file is saved, and update the command line as follows</p>
</div></blockquote>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m cansig.run.pipeline data/pipeline-tutorial/data.hdf5
                                --batch batch
                                --gene-sets h.all.v<span class="o">{</span>current_version<span class="o">}</span>.symbols.gmt <span class="se">\</span>
                                --dimensions <span class="m">4</span> <span class="m">6</span> --model-runs <span class="m">1</span> <span class="se">\</span>
                                --clusters <span class="m">2</span> <span class="m">3</span> <span class="m">5</span>  --cluster-runs <span class="m">1</span> <span class="se">\</span>
                                --output tutorial-output
</pre></div>
</div>
</div>
<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<p>Fill in the TODO items above.</p>
</div>
<p>The results will be available in the generated directory <code class="docutils literal notranslate"><span class="pre">tutorial-output/</span></code> in a few minutes.</p>
<p>Let’s analyse its structure:</p>
<ul class="simple">
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">integration/</span></code>: each model has a separate directory representing the latent representations inferred by the batch correction/dimension reduction method</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/integration-settings.json</span></code>: summary of the parameters used for the integration method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/latent-representations.csv</span></code>: coordinates in the latent space inferred by the integration method</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">postprocessing/</span></code>: each clustering configuration on each model has a separate directory with the postprocessing information saved</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/cluster-labels.csv</span></code>: cluster assignments for this run</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/cluster-settings.json</span></code>: summary of the parameters used for the clustering step</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/gsea-dataframe.csv</span></code>: results of GSEA run in this setting (full table, including non-significant pathways and negative enrichment score pathways)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/gsea-settings.json</span></code>: summary of the parameters used for the GSEA run</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/integration-settings.json</span></code>: summary of the parameters used for the integration method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/latent-space-dimred.png</span></code>: scatter plot of the PCA on the latent space colored according to batch and clustering (see advanced tutorial for more details)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/cells-score-denovo-signature.csv</span></code>: scores for each cell for the de novo found signatures for this run</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/denovo-signature-correlation.csv</span></code>: the pearson correlation between de novo found signatures</p></li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">{rundir}/signatures/</span></code>: directory containing the de novo signatures as the results of the differential gene expression analysis</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">signature_cl{CLUSTER}.csv</span></code>: results of the differential gene expression analysis for cluster CLUSTER, ranked according to z-score.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">heatmap.pdf</span></code>: heatmap summarizing Normalized Enrichment Score of the pathways which seem to be differentially expressed over all runs indicated</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">representative-directory.txt</span></code>: simple text file with the name of the post processing directory that can be used as a “representative directory”</p></li>
</ul>
<p>While we covered the most basic usage of the pipeline, more information can be obtained by running</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m cansig.run.pipeline --help
</pre></div>
</div>
<p>or by consulting the <a class="reference internal" href="pipeline-advanced.html#pipeline-advanced"><span class="std std-ref">Advanced pipeline usage</span></a>.
Options include running your own integration model, controlling the plotting utilities, performing differential CNV analysis</p>
</section>
<section id="interpreting-the-results">
<h3>Interpreting the results<a class="headerlink" href="#interpreting-the-results" title="Permalink to this heading">#</a></h3>
<div class="admonition-todo admonition" id="id4">
<p class="admonition-title">Todo</p>
<p>Basic discussion how to interpret the results.</p>
</div>
<p>For more advanced analysis, including drawing biological insights, see <a class="reference internal" href="interpretation.html#interpretation"><span class="std std-ref">Data interpretation</span></a>.</p>
</section>
</section>
<section id="tutorials">
<h2>Tutorials<a class="headerlink" href="#tutorials" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>To learn more about the pipeline (parallelization, using custom models, plotting), see <a class="reference internal" href="pipeline-advanced.html#pipeline-advanced"><span class="std std-ref">Advanced pipeline usage</span></a>.</p></li>
<li><p>For a tutorial how to interpret the results from the biological perspective, see <a class="reference internal" href="interpretation.html#interpretation"><span class="std std-ref">Data interpretation</span></a>.</p></li>
<li><p>To learn about the preprocessing module (used to prepare the raw data into the .h5ad format), see the <a class="reference internal" href="preprocessing.html#preprocessing"><span class="std std-ref">Preprocessing</span></a>.</p></li>
<li><p>To learn how to format already preprocessed .csv files into .h5ad files, see <a class="reference internal" href="formatting-data.html#formatting"><span class="std std-ref">Input data format</span></a>.</p></li>
<li><p>To learn how to run CanSig in a python script/jupyter notebook, see <a class="reference internal" href="cansig-for-python-coders.html#coders"><span class="std std-ref">Cansig for python coders</span></a>.</p></li>
<li><p>To ensure smooth running of CanSig, check out <a class="reference internal" href="troubleshooting.html#troubleshooting"><span class="std std-ref">the checklist and troubleshooting page.</span></a>.</p></li>
</ul>
</section>
<section id="contributing">
<h2>Contributing<a class="headerlink" href="#contributing" title="Permalink to this heading">#</a></h2>
<p>For the contribution guide and instructions for new developers, see <a class="reference internal" href="contributing.html#contribution-guide"><span class="std std-ref">Contributing</span></a>.</p>
</section>
</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='right-next' id="next-link" href="preprocessing.html" title="next page">
  <div class="prev-next-info">
      <p class="prev-next-subtitle">next</p>
      <p class="prev-next-title">Preprocessing</p>
  </div>
  <i class="fa-solid fa-angle-right"></i>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
<div class="tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
</div>
<nav id="bd-toc-nav" class="page-toc">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#navigation">
   Navigation
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#getting-started">
   Getting started
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#installation">
     Installation
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example-dataset">
     Example dataset
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#running-the-analysis">
     Running the analysis
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#interpreting-the-results">
     Interpreting the results
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tutorials">
   Tutorials
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributing">
   Contributing
  </a>
 </li>
</ul>

</nav>
</div>

<div class="toc-item">
  
<div id="searchbox"></div>
</div>

<div class="toc-item">
  
</div>

<div class="toc-item">
  
<div class="tocsection sourcelink">
    <a href="_sources/index.rst.txt">
        <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
</div>

</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2022 CanSig team.<br>

</p>

  </div>
  
  <div class="footer-item">
    <p class="theme-version">
    Built with the
    <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">
        PyData Sphinx Theme
    </a>
    0.12.0.
</p>
  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.3.0.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>