
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>CanSig: discovering cancer signatures &#8212; CanSig 0.0.1 documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Preprocessing tutorial" href="preprocessing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
    
    <div class="container-fluid" id="banner"></div>

    
    <nav class="navbar navbar-light navbar-expand-lg bg-light fixed-top bd-navbar" id="navbar-main"><div class="container-xl">

  <div id="navbar-start">
    
    
<a class="navbar-brand" href="#">
<p class="title">CanSig</p>
</a>

    
  </div>

  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar-collapsible" aria-controls="navbar-collapsible" aria-expanded="false" aria-label="Toggle navigation">
    <span class="navbar-toggler-icon"></span>
  </button>

  
  <div id="navbar-collapsible" class="col-lg-9 collapse navbar-collapse">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <ul id="navbar-main-elements" class="navbar-nav">
    <li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="preprocessing.html">
  Preprocessing tutorial
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="pipeline-advanced.html">
  Advanced pipeline usage
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="interpretation.html">
  Advanced interpretation tutorial
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="contributing.html">
  Contribution guide
 </a>
</li>

<li class="toctree-l1 nav-item">
 <a class="reference internal nav-link" href="source/modules.html">
  cansig
 </a>
</li>

    
</ul>
      </div>
      
    </div>

    <div id="navbar-end">
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          <a class="nav-link" href="https://github.com/BoevaLab/CanSig" rel="noopener" target="_blank" title="GitHub"><span><i class="fab fa-github-square"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
      </ul>
      </div>
      
    </div>
  </div>
</div>
    </nav>
    

    <div class="container-xl">
      <div class="row">
          
            
            <!-- Only show if we have sidebars configured, else just a small margin  -->
            <div class="col-12 col-md-3 bd-sidebar">
              <div class="sidebar-start-items"><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  <div class="bd-toc-item active">
    
  </div>
</nav>
              </div>
              <div class="sidebar-end-items">
              </div>
            </div>
            
          

          
          <div class="d-none d-xl-block col-xl-2 bd-toc">
            
              
              <div class="toc-item">
                
<div class="tocsection onthispage mt-5 pt-1 pb-3">
    <i class="fas fa-list"></i> On this page
</div>

<nav id="bd-toc-nav">
    <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#getting-started">
   Getting started
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#installation">
     Installation
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#example-dataset">
     Example dataset
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#running-the-analysis">
     Running the analysis
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#interpreting-the-results">
     Interpreting the results
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#tutorials">
   Tutorials
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#contributing">
   Contributing
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#indices-and-tables">
   Indices and tables
  </a>
 </li>
</ul>

</nav>
              </div>
              
              <div class="toc-item">
                
              </div>
              
            
          </div>
          

          
          
            
          
          <main class="col-12 col-md-9 col-xl-7 py-md-5 pl-md-5 pr-md-4 bd-content" role="main">
              
              <div>
                
  <section id="cansig-discovering-cancer-signatures">
<h1>CanSig: discovering cancer signatures<a class="headerlink" href="#cansig-discovering-cancer-signatures" title="Permalink to this heading">#</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This package is currently <em>not</em> ready to be used. We hope it will be available to public use in June.</p>
</div>
<p>Human tumors are highly heterogeneous in their cell composition; specifically, they exhibit heterogeneity in transcriptional states of malignant cells, as has been recently discovered through single-cell RNA sequencing (scRNA-seq). Distinct states of malignant cells have been linked to variability in tumorigenic properties and resistance to anti-cancer treatment. Despite the fact that scRNA-seq data contain all necessary information to uncover shared transcriptional states of malignant cells in tumors, jointly analyzing cells from multiple cancer patients comes with its set of challenges including batch correction and accounting for patient-specific genetic background driving differences between gene expression vectors. We propose CanSig, an easy-to-use approach designed to discover known and de novo shared signatures in cancer single cells. CanSig preprocesses, integrates and analyzes scRNA-seq data to provide new signatures of shared transcriptional states and links these states to known pathways.</p>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>Add a picture of the pipeline here with the figure caption as in the preprint.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A preprint describing the pipeline and case studies is <a class="reference external" href="https://doi.org/10.1101/2022.04.14.488324">now available</a>.</p>
</div>
<section id="getting-started">
<h2>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this heading">#</a></h2>
<p>This tutorial assumes you have access to a Unix terminal (e.g., Linux, BSD, MacOS). If you use Windows, install <a class="reference external" href="https://docs.microsoft.com/en-us/windows/wsl/install">Windows Subsystem for Linux</a> first.</p>
<p>When the Unix terminal is ready, you will need to <a class="reference external" href="https://wiki.python.org/moin/BeginnersGuide">install Python</a>. One popular solution is <a class="reference external" href="https://www.anaconda.com/">Anaconda</a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>CanSig requires Python 3.8 (or a most recent version).</p>
</div>
<section id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Permalink to this heading">#</a></h3>
<p>Install the package:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ pip install cansig
</pre></div>
</div>
</section>
<section id="example-dataset">
<h3>Example dataset<a class="headerlink" href="#example-dataset" title="Permalink to this heading">#</a></h3>
<p>To demonstrate the pipeline usage, we will analyse a small esophageal dataset, which has been already preprocessed (for more information about the dataset and the data curation, see <a class="reference internal" href="preprocessing.html#preprocessing"><span class="std std-ref">Preprocessing tutorial</span></a>).</p>
<p>We will download it using an auxiliary script:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m cansig.run.download PIPELINE
</pre></div>
</div>
<p>You should see a new directory <code class="docutils literal notranslate"><span class="pre">data</span></code> with a file <code class="docutils literal notranslate"><span class="pre">pipeline-tutorial.h5ad</span></code>.</p>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>Add the description of this dataset and UMAP visualisations.</p>
</div>
</section>
<section id="running-the-analysis">
<h3>Running the analysis<a class="headerlink" href="#running-the-analysis" title="Permalink to this heading">#</a></h3>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you are familiar with Python, you may prefer running the analysis in a custom manner. See the <a class="reference internal" href="pipeline-advanced.html#pipeline-advanced"><span class="std std-ref">Advanced pipeline usage</span></a>.</p>
</div>
<p>We will run the analysis pipeline with a single command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m cansig.run.pipeline data/pipeline-tutorial/data.hdf5
                                --batch batch
                                --gene-sets MSigDB_Hallmark_2020 <span class="se">\</span>
                                --dimensions <span class="m">4</span> <span class="m">6</span> --model-runs <span class="m">1</span> <span class="se">\</span>
                                --clusters <span class="m">2</span> <span class="m">3</span> <span class="m">5</span>  --cluster-runs <span class="m">1</span> <span class="se">\</span>
                                --output tutorial-output
</pre></div>
</div>
<p>This command launched the training of several models, used for data integration (batch correction and dimension reduction).
We marked that the batch column in the dataset is named <code class="docutils literal notranslate"><span class="pre">batch</span></code> and we will reduce the dimension either to 4 or 6 (<code class="docutils literal notranslate"><span class="pre">--dimensions</span> <span class="pre">4</span> <span class="pre">6</span></code>).
For each of these dimensionalities, we do a single run (<code class="docutils literal notranslate"><span class="pre">--model-runs</span> <span class="pre">1</span></code>). Increasing this number will train more models, differing in random seeds.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In case of additional continuous or discrete variables, which may confound the expression (e.g., the sequencing technology used), one could specify it via the <code class="docutils literal notranslate"><span class="pre">--continuous-covariates</span></code> and <code class="docutils literal notranslate"><span class="pre">--discrete-covariates</span></code> flags.</p>
<p>For more information, run</p>
<p><code class="docutils literal notranslate"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">-m</span> <span class="pre">cansig.run.pipeline</span> <span class="pre">--help</span></code></p>
</div>
<p>Then, for each of the four models, we apply clustering into 2, 3, or 5 communities (<code class="docutils literal notranslate"><span class="pre">--clusters</span> <span class="pre">2</span> <span class="pre">3</span> <span class="pre">5</span></code>).
Again, one can repeat this steps, using different random seeds, by increasing <code class="docutils literal notranslate"><span class="pre">--cluster-runs</span></code>.</p>
<p>We specified the database to be used (<code class="docutils literal notranslate"><span class="pre">--gene-sets</span> <span class="pre">MSigDB_Hallmark_2020</span></code>).
This is one of many databases available online. Alternatively, one can provide a path to the GMT file.</p>
<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<p>Fill in the TODO items above.</p>
</div>
<p>The results will be available in the generated directory <code class="docutils literal notranslate"><span class="pre">tutorial-output/</span></code> in a few minutes.</p>
<p>Letâ€™s analyse its structure:</p>
<ul class="simple">
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">integration/</span></code>: each model has a separate directory representing the latent representations inferred by the batch correction/dimension reduction method</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/integration-settings.json</span></code>: summary of the parameters used for the integration method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/latent-representations.csv</span></code>: coordinates in the latent space inferred by the integration method</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">postprocessing/</span></code>: each clustering configuration on each model has a separate directory with the postprocessing information saved</dt><dd><ul>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/cluster-labels.csv</span></code>: cluster assignments for this run</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/cluster-settings.json</span></code>: summary of the parameters used for the clustering step</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/gsea-dataframe.csv</span></code>: results of GSEA run in this setting (full table, including non-significant pathways and negative enrichment score pathways)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/gsea-settings.json</span></code>: summary of the parameters used for the GSEA run</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/integration-settings.json</span></code>: summary of the parameters used for the integration method</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">{rundir}/latent_space_dimred.png</span></code>: by default, scatter plot of the PCA on the latent space colored according to batch and clustering (see advanced tutorial for more details)</p></li>
</ul>
</dd>
</dl>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">heatmap.pdf</span></code>: heatmap summarizing Normalized Enrichment Score of the pathways which seem to be differentially expressed over all runs indicated</p></li>
</ul>
<p>While we covered the most basic usage of the pipeline, more information can be obtained by running</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ python -m cansig.run.pipeline --help
</pre></div>
</div>
<p>or by consulting the <a class="reference internal" href="pipeline-advanced.html#pipeline-advanced"><span class="std std-ref">Advanced pipeline usage</span></a>.
Options include running your own integration model or controlling the plotting utilities.</p>
</section>
<section id="interpreting-the-results">
<h3>Interpreting the results<a class="headerlink" href="#interpreting-the-results" title="Permalink to this heading">#</a></h3>
<div class="admonition-todo admonition" id="id4">
<p class="admonition-title">Todo</p>
<p>Basic discussion how to interpret the results.</p>
</div>
<p>For more advanced analysis, including drawing biological insights, see <a class="reference internal" href="interpretation.html#interpretation"><span class="std std-ref">Advanced interpretation tutorial</span></a>.</p>
</section>
</section>
<section id="tutorials">
<h2>Tutorials<a class="headerlink" href="#tutorials" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>To learn more about the pipeline (parallelization, using custom models, plotting), see <a class="reference internal" href="pipeline-advanced.html#pipeline-advanced"><span class="std std-ref">Advanced pipeline usage</span></a>.</p></li>
<li><p>For a tutorial how to interpret the results from the biological perspective, see <a class="reference internal" href="interpretation.html#interpretation"><span class="std std-ref">Advanced interpretation tutorial</span></a>.</p></li>
<li><p>To learn about the preprocessing module (used to prepare the raw data into the HDF5 format), see the <a class="reference internal" href="preprocessing.html#preprocessing"><span class="std std-ref">Preprocessing tutorial</span></a>.</p></li>
</ul>
</section>
<section id="contributing">
<h2>Contributing<a class="headerlink" href="#contributing" title="Permalink to this heading">#</a></h2>
<p>For the contribution guide and instructions for new developers, see <a class="reference internal" href="contributing.html#contribution-guide"><span class="std std-ref">Contribution guide</span></a>.</p>
</section>
<section id="indices-and-tables">
<h2>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this heading">#</a></h2>
<div class="toctree-wrapper compound">
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="preprocessing.html">Preprocessing tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#loading-the-data">Loading the data</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#id3">Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#outputs">Outputs</a></li>
<li class="toctree-l2"><a class="reference internal" href="preprocessing.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pipeline-advanced.html">Advanced pipeline usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#using-custom-models">Using custom models</a></li>
<li class="toctree-l2"><a class="reference internal" href="pipeline-advanced.html#using-the-plotting-utilities">Using the plotting utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="interpretation.html">Advanced interpretation tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contribution guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#submitting-an-issue">Submitting an issue</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#submitting-a-pull-request">Submitting a Pull Request</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#for-new-developers">For new developers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="source/modules.html">cansig</a><ul>
<li class="toctree-l2"><a class="reference internal" href="source/cansig.html">cansig package</a></li>
</ul>
</li>
</ul>
</div>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>
</section>


              </div>
              
              
              <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='right-next' id="next-link" href="preprocessing.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Preprocessing tutorial</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
              
          </main>
          

      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>
<footer class="footer mt-5 mt-md-0">
  <div class="container">
    
    <div class="footer-item">
      <p class="copyright">
    &copy; Copyright 2022 CanSig team.<br>
</p>
    </div>
    
    <div class="footer-item">
      <p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.0.1.<br>
</p>
    </div>
    
  </div>
</footer>
  </body>
</html>